File: app/src/main/java/com/example/degoogle/ui/info/AppInfoViewModel.java
package com.example.degoogle.ui.info;

import android.app.Application;
import android.util.Log;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.work.Data;
import androidx.work.ExistingWorkPolicy;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkManager;

import com.example.degoogle.data.entities.InstalledPackages;
import com.example.degoogle.model.Product;
import com.example.degoogle.repositories.PackageRepository;
import com.example.degoogle.workers.DownloadWorker;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.io.File;
import java.util.List;

public class AppInfoViewModel extends AndroidViewModel{
    private PackageRepository packageRepository;
    FirebaseFirestore db = FirebaseFirestore.getInstance();
    MutableLiveData<Product> packageInfo = new MutableLiveData<>();
    private WorkManager workManager;
    private final LiveData<List<InstalledPackages>> installedPackages;

    public AppInfoViewModel(Application application) {
        super(application);
        packageRepository = new PackageRepository(application);
        installedPackages = packageRepository.getAllPackages();
        workManager = WorkManager.getInstance(application);
    }

    public LiveData<List<InstalledPackages>> getAllPackages() {
        return installedPackages;
    }
    public void insert(InstalledPackages installedPackages) { packageRepository.insert(installedPackages);
        Log.d("AppInfoViewModel: ", "inserted" + installedPackages.getPackageName() + " " + installedPackages.getPackageVersion());
        Log.d("AppInfoViewModel: ", "table content" + packageRepository.getAllPackages().getValue());
    }


    public void getDataFromFirebase(String id) {

        db.collection("apps")
                .document(id)
                .get()
                .addOnCompleteListener(task -> {
                    DocumentSnapshot document = task.getResult();
//                    packageInfo.setValue(document.toObject(AppInfoModel.class));
                    packageInfo.postValue(document.toObject(Product.class));
                    Log.d("ViewModelAppInfo", "getDataFromFirebase: " + packageInfo.getValue());
                    Log.d("ViewModelAppInfo", "getDataFromFirebase: " + document.toObject(Product.class));
                 });


    }
    public Product getModel(){return packageInfo.getValue();}
    public LiveData<Product> getPackageInfo() {return packageInfo;}

    public void installApk(String packageName, File file) {
        Data data = new Data.Builder().putString("file", file.toURI().toString()).build();
        OneTimeWorkRequest.Builder downloadWorker = new OneTimeWorkRequest.Builder(DownloadWorker.class).setInputData(data);
        workManager.enqueueUniqueWork("packageInstall", ExistingWorkPolicy.KEEP ,downloadWorker.build());
    }
}

File: app/src/main/java/com/example/degoogle/ui/info/AppInfoViewModel.kt
package com.example.degoogle.ui.info

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.example.degoogle.data.entities.InstalledPackages
import com.example.degoogle.data.PackagesDao
import kotlinx.coroutines.launch

class AppInfoViewModel(private val packagesDao: PackagesDao) : ViewModel() {
    private fun insertItem(item: InstalledPackages){
        viewModelScope.launch {
            packagesDao.insert(item)
        }

    }
    private fun getNewItem(packageName: String, version: String): InstalledPackages {
        return InstalledPackages(
            appPackage = packageName,
            version = version
        )
    }
    fun addNewItem(packageName: String, version: String){
        val newItem = getNewItem(packageName, version)
        insertItem(newItem)
        }
    fun isEntryValid(packageName: String, version: String): Boolean {
        if (packageName.isBlank() || version.isBlank()) {
            return false
        }
        return true
    }
}
class AppInfoViewModelFactory(private val packagesDao: PackagesDao) : ViewModelProvider.Factory{
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(AppInfoViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return AppInfoViewModel(packagesDao) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}